knative-operator/pkg/common/service_monitor.go

```diff
@@ -54,23 +80,136 @@ func SetupServiceMonitor(ctx context.Context, cfg *rest.Config, metricsPort int3
+	monitor := &monitoringv1.ServiceMonitor{}
+	var SchemeGroupVersion = schema.GroupVersion{Group: "monitoring.coreos.com", Version: "v1"}
+	scheme.Scheme.AddKnownTypes(SchemeGroupVersion, monitor)
+	if err := manifest.Apply(); err != nil {
 		return err
 	}
-	// To generate metrics in other namespaces, add the values below.
-	ns := []string{operatorNs}
-	// Generate and serve custom resource specific metrics.
-	err = kubemetrics.GenerateAndServeCRMetrics(cfg, ns, filteredGVK, metricsHost, operatorMetricsPort)
+	return nil
+}
+
+func SetupSourceServiceMonitor(client client.Client, instance *appsv1.Deployment) error {
+	monitor := &monitoringv1.ServiceMonitor{}
+	gv := schema.GroupVersion{Group: "monitoring.coreos.com", Version: "v1"}
+	scheme.Scheme.AddKnownTypes(gv, monitor)
+	labels := instance.Spec.Selector.MatchLabels
+
+	clientOptions := mf.UseClient(mfclient.NewClient(client))
+	// create service for the deployment
+	manifest, err := mf.NewManifest(getMonitorPath(TestSourceServicePath, EventingSourcePath), clientOptions)
+	if err != nil {
+		return fmt.Errorf("unable to parse source service manifest: %w", err)
+	}
+	transforms := []mf.Transformer{updateService(labels, instance.Name), mf.InjectOwner(instance), mf.InjectNamespace(instance.Namespace)}
+	if manifest, err = manifest.Transform(transforms...); err != nil {
+		return fmt.Errorf("unable to transform source service manifest: %w", err)
+	}
+	if err := manifest.Apply(); err != nil {
+		return err
+	}
+
+	// get service back, needed for the UID and setting owner refs
+	srv := &v1.Service{}
+	if err := client.Get(context.TODO(), types.NamespacedName{Name: instance.Name, Namespace: instance.Namespace}, srv); err != nil {
+		return err
+	}
+	// create service monitor for source
+	manifest, err = mf.NewManifest(getMonitorPath(TestSourceServiceMonitorPath, EventingSourceServiceMonitorPath), clientOptions)
 	if err != nil {
+		return fmt.Errorf("unable to parse source service monitor manifest: %w", err)
+	}
+	transforms = []mf.Transformer{updateServiceMonitor(labels, instance.Name), mf.InjectOwner(srv), mf.InjectNamespace(instance.Namespace)}
+	if manifest, err = manifest.Transform(transforms...); err != nil {
+		return fmt.Errorf("unable to transform source service monitor manifest: %w", err)
+	}
+	if err := manifest.Apply(); err != nil {
 		return err
 	}
 	return nil
 }
+
+func getMonitorPath(envVar string, defaultVal string) string {
+	path := os.Getenv(envVar)
+	if path == "" {
+		return defaultVal
+	}
+	return path
+}
+
+func updateService(labels map[string]string, depName string) mf.Transformer {
+	return func(resource *unstructured.Unstructured) error {
+		if resource.GetKind() != "Service" {
+			return nil
+		}
+		var svc = &v1.Service{}
+		if err := scheme.Scheme.Convert(resource, svc, nil); err != nil {
+			return err
+		}
+		svc.Name = depName
+		svc.Labels = kmeta.CopyMap(labels)
+		svc.Spec.Selector = kmeta.CopyMap(labels)
+		svc.Labels["name"] = svc.Name
```

Comment:
We need the name so that the service monitor can select it, I follow what the monitoring package does which passed the service labels to the service monitor generation function.